### PR TITLE
[16.0][ADD] account_analytic_relation_filter

### PR DESCRIPTION
**Pull request by AungKoKoLin1997**
This module introduces a filter for the analytic_distribution field by establishing a relation between analytic accounts.

@qrtl QT5540

### CODE REVIEW

File: analytic_distribution_filter/models/analytic_account_line.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    def create(self, vals):
+        res = super().create(vals)
+        for record in res:
+            record._update_reciprocal_relations()
+        return res
+
+    def write(self, vals):
+        for record in self:
+            old_ids = record.account_ids.ids
+            super().write(vals)
```
Comment (kanda999) [inline analytic_distribution_filter/models/analytic_account_line.py]: `super().write(vals)`
Should we have this in the loop?

Comment (AungKoKoLin1997) [reply analytic_distribution_filter/models/analytic_account_line.py]: I called super() inside of loop after getting old_ids but my previous code has logic problem if there are multiple records are updated at once. Updated the logic.

**Review by kanda999 — APPROVED** [review]
Code and functional review

File: analytic_distribution_filter/readme/CONFIGURE.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
+To define related accounts (used by the filter)
+
+1.  Go to **Invoicing → Configuration → Analytic Distribution
+    Accounts**.
```
Comment (yostashiro) [inline analytic_distribution_filter/readme/CONFIGURE.md]:
```suggestion
1.  Go to *Invoicing → Configuration → Analytic Distribution
    Accounts*.
```

File: analytic_distribution_filter/readme/CONFIGURE.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    - If no related accounts are set, the widget falls back to standard
+      (unfiltered) behavior.
+
+## Example
+
+You have three plans, each with two accounts:
+
+- **Plan A**: Account A, Account B
+- **Plan B**: Account C, Account D
+- **Plan C**: Account E, Account F
```
Comment (yostashiro) [inline analytic_distribution_filter/readme/CONFIGURE.md]: To avoid confusion.
```suggestion
- **Plan 1**: Account A, Account B
- **Plan 2**: Account C, Account D
- **Plan 3**: Account E, Account F
```

File: analytic_distribution_filter/models/analytic_account.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+from odoo import fields, models
+
+
+class AccountAnalyticAccount(models.Model):
+    _inherit = "account.analytic.account"
+
+    account_line_ids = fields.One2many("analytic.account.line", "account_id")
+
+    def get_related_account_ids(self):
+        """Return related analytic account for distribution filtering."""
```
Comment (yostashiro) [inline analytic_distribution_filter/models/analytic_account.py]:
```suggestion
        """Return related analytic accounts for distribution filtering."""
```

File: analytic_distribution_filter/models/analytic_account_line.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
+# Copyright 2025 Quartile (https://www.quartile.co)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import Command, api, fields, models
+
+
+class AnalyticAccountLine(models.Model):
+    _name = "analytic.account.line"
```
Comment (yostashiro) [inline analytic_distribution_filter/models/analytic_account_line.py]: To better represent what it is.
```suggestion
    _name = "analytic.account.relation.line"
    _description = "Analytic Account Relation Line"
```

File: analytic_distribution_filter/models/analytic_account.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
+# Copyright 2025 Quartile (https://www.quartile.co)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class AccountAnalyticAccount(models.Model):
+    _inherit = "account.analytic.account"
+
+    account_line_ids = fields.One2many("analytic.account.line", "account_id")
```
Comment (yostashiro) [inline analytic_distribution_filter/models/analytic_account.py]:
```suggestion
    relation_line_ids = fields.One2many("analytic.account.relation.line", "account_id")
```

File: analytic_distribution_filter/models/analytic_account_line.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+                    if line.account_id.id not in reciprocal.account_ids.ids:
+                        reciprocal.account_ids = [Command.link(line.account_id.id)]
+                else:
+                    self.create(
+                        {
+                            "account_id": account_id,
+                            "plan_id": line.account_id.plan_id.id,
+                            "account_ids": [Command.set([line.account_id.id])],
+                        }
+                    )
```
Comment (yostashiro) [inline analytic_distribution_filter/models/analytic_account_line.py]: To improve performance.
```suggestion
            existing_recip_lines = self._get_reciprocal_lines(added_ids, line.account_id.plan_id.id)
            missing_ids = added_ids - set(existing_recip_lines.account_id.ids)
            if missing_ids:
                self.create(
                    [
                        {
                            "account_id": aid,
                            "plan_id": line.account_id.plan_id.id,
                            "related_account_ids": [Command.set([line.account_id.id])],
                        }
                        for aid in missing_ids
                    ]
                )
            if existing_recip_lines:
                    existing.write({"related_account_ids": [Command.link(line.account_id.id)]})
```

File: analytic_distribution_filter/models/analytic_account_line.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        self.account_ids = False
+
+    def _search_reciprocal(self, account_id, plan_id):
+        return self.search(
+            [
+                ("account_id", "=", account_id),
+                ("plan_id", "=", plan_id),
+            ],
+            limit=1,
+        )
```
Comment (yostashiro) [inline analytic_distribution_filter/models/analytic_account_line.py]:
```suggestion
    def _get_reciprocal_lines(self, account_ids, plan_id):
        return self.search(
            [("account_id", "in", account_ids), ("plan_id", "=", plan_id)],
        )
```

File: analytic_distribution_filter/models/analytic_account_line.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+                        }
+                    )
+            for account_id in removed_ids:
+                reciprocal = self._search_reciprocal(
+                    account_id, line.account_id.plan_id.id
+                )
+                if reciprocal and line.account_id.id in reciprocal.account_ids.ids:
+                    reciprocal.account_ids = [Command.unlink(line.account_id.id)]
+                    if not reciprocal.account_ids:
+                        reciprocal.unlink()
```
Comment (yostashiro) [inline analytic_distribution_filter/models/analytic_account_line.py]:
```suggestion
            if remove_ids:
            recip_lines = self._get_reciprocal_lines(removed_ids, line.account_id.plan_id.id)
            recip_lines.write({"account_ids": [Command.unlink(rec.account_id.id)]})
            recip_lines.filtered(lambda x: not x.account_ids).unlink()
```

File: analytic_distribution_filter/models/analytic_account_line.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+from odoo import Command, api, fields, models
+
+
+class AnalyticAccountLine(models.Model):
+    _name = "analytic.account.line"
+
+    account_id = fields.Many2one("account.analytic.account", ondelete="cascade")
+    plan_id = fields.Many2one("account.analytic.plan", required=True)
+    account_ids = fields.Many2many(
+        "account.analytic.account", string="Related Accounts", required=True
```
Comment (yostashiro) [inline analytic_distribution_filter/models/analytic_account_line.py]: I think we can make the field non-required.
```suggestion
        "account.analytic.account", string="Related Accounts",
```

File: analytic_distribution_filter/models/analytic_account_line.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    def unlink(self):
+        for record in self:
+            for acc in record.account_ids:
+                reciprocal = self._search_reciprocal(
+                    acc.id, record.account_id.plan_id.id
+                )
+                if reciprocal and record.account_id.id in reciprocal.account_ids.ids:
+                    reciprocal.account_ids = [Command.unlink(record.account_id.id)]
+                    if not reciprocal.account_ids:
+                        reciprocal.unlink()
```
Comment (yostashiro) [inline analytic_distribution_filter/models/analytic_account_line.py]: Update this using _get_reciprocal_lines().

**Review by yostashiro — COMMENTED** [review]
Name suggestion: account_analytic_relation_filter

File: account_analytic_relation_filter/static/tests/test_analytic_distribution.esm.js [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+                            if (args[0][0] === 1) {
+                                return [2, 3];
+                            }
+                            return [4, 5];
+                        }
+                        return [];
+                    },
+                },
+                relatedAccountIDs: [],
+                existingAnalyticAccountIDs: [1],
```
Comment (yostashiro) [inline account_analytic_relation_filter/static/tests/test_analytic_distribution.esm.js]: Non-blocking, but it will make it easier to understand the flow if we move this before the onSelect() test.

**Review by yostashiro — APPROVED** [review]
Code review.

File: account_analytic_relation_filter/static/tests/test_analytic_distribution.esm.js (hunk 12 above)
Comment (AungKoKoLin1997) [reply account_analytic_relation_filter/static/tests/test_analytic_distribution.esm.js]: Done!

**Comment by AungKoKoLin1997** [comment]
@yostashiro @kanda999 I make some adjustments to work the filter even when we select the accounts from the search more... as well and some JS code improvements.

**Comment by yostashiro** [comment]
Filter doesn't seem to be applied in some situations.

"Internal" department is linked to two projects.
image

"Internal" department has been selected for a record and **the record has been saved**.
image

No filter is applied to projects (which is not as expected).
image

Once I open the search modal and close it, the filter is applied correctly.
image

**Comment by AungKoKoLin1997** [comment]
> Filter doesn't seem to be applied in some situations.

@yostashiro I fixed it.

**Comment by yostashiro** [comment]
Not sure where it's coming from but I get a client error when I try to select an analytic account for saved records.
image

**Review by yostashiro — APPROVED** [review]
LGTM. The JS error I reported doesn't seem to reproduce now.

**Review by kanda999 — APPROVED** [review]
Functional review: it works as expected

**Comment by AungKoKoLin1997** [comment]
@oca/account-analytic-maintainers
Could you please take a look at this PR?

**Comment by AungKoKoLin1997** [comment]
My latest commit is to fix the crash that happened because our patch was still awaiting an RPC result (orm.call) after the component had already been destroyed (for example, when a UserError interrupted the form creation). When that happens, OWL raises “Component is destroyed”.

**Review by kanda999 — APPROVED** [review]
Functional review: It’s still working on the production server.

**Review by marcos-mendez — APPROVED** [review]
**Automated Review -- Tests Passed**
All tests for `account_analytic_relation_filter` passed successfully on Odoo 16.0.
Environment: Minikube + K8s Job + oca-ci/py3.10-odoo16.0
---
### Test Coverage Suggestions
### Coverage Gaps

The existing tests likely cover basic functionality, but there are several areas with potential gaps:

1. **Reciprocal relation updates** in `analytic_account_relation_line.py`:
   - The `_update_reciprocal_relations` method has complex logic for maintaining bidirectional links.
   - No tests cover scenarios where:
     - Adding/removing accounts causes cascading updates
     - Unlinking accounts removes reciprocal entries correctly
     - Multiple accounts are linked/unlinked simultaneously

2. **Domain computation edge cases**:
   - `_compute_plan_id_domain` may not be tested with:
     - Empty relation lines
     - Accounts with no related plans
     - Accounts with many related plans

3. **Filtering logic in `get_related_account_ids`**:
   - No test ensures that accounts not in related plans are included correctly
   - Edge case: when `relation_line_ids` is empty, fallback behavior is not explicitly tested

4. **JavaScript component behavior**:
   - No QUnit tests for frontend behavior
   - No tests for interaction between JS and Odoo's `analytic_distribution` widget

### Suggested Test Cases

```python
def test_compute_plan_id_domain_empty_relations(self):
    """Test plan_id_domain when no related accounts exist."""
    # Create analytic account with no relation lines
    account = self.env['account.analytic.account'].create({'name': 'Test'})
    line = self.env['analytic.account.relation.line'].create({
        'account_id': account.id,
        'plan_id': self.env['account.analytic.plan'].search([], limit=1).id,
    })
    # Ensure domain is computed correctly for empty related accounts
    line._compute_plan_id_domain()
    self.assertTrue(line.plan_id_domain)

def test_update_reciprocal_relations_add_remove(self):
    """Test that adding and removing related accounts updates reciprocal lines."""
    # Setup: Create two accounts and link them
    account1 = self.env['account.analytic.account'].create({'name': 'A1'})
    account2 = self.env['account.analytic.account'].create({'name': 'A2'})
    plan = self.env['account.analytic.plan'].search([], limit=1)
    # Create relation line
    relation = self.env['analytic.account.relation.line'].create({
        'account_id': account1.id,
        'plan_id': plan.id,
        'account_ids': [(4, account2.id)],
    })
    # Add another account to the relation
    relation.account_ids = [(4, self.env['account.analytic.account'].create({'name': 'A3'}).id)]
    # Verify reciprocal lines were updated
    self.assertTrue(relation.account_ids)

def test_get_related_account_ids_empty_relations(self):
    """Test get_related_account_ids fallback behavior when no relations exist."""
    account = self.env['account.analytic.account'].create({'name': 'Test'})
    related_ids = account.get_related_account_ids()
    # Should return all accounts in related plans or all accounts if no relations
    self.assertIsInstance(related_ids, list)
```

### Codecov Risk

- `get_related_account_ids` in `analytic_account.py`: If not tested with edge cases (no relations, empty plans), it could reduce coverage.
- `_update_reciprocal_relations` in `analytic_account_relation_line.py`: Complex logic with multiple branches; likely to be a coverage risk without tests for all code paths.
- `_compute_plan_id_domain`: Edge case handling for empty records not explicitly tested.

These methods are critical to the module's behavior and should be thoroughly tested.
---
_Automated review by [OCA Neural Reviewer](https://git.pop.coop/pop/oca-reviewer) + qwen3-coder:30b_
